### ISSUE TITLE
`elided_lifetime_in_path` triggers for the `format!` macro

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#48385]
I was trying to use `deny(elided_lifetime_in_path)` and encountered the problem that [this doesn't work](https://play.rust-lang.org/?gist=4b2834a15051c990e5dab02c85732cc4&version=nightly):

```rust
#![feature(nll)]
#![deny(elided_lifetime_in_path)]

fn main() {
    format!("foo {}", 22)
}
```

I get 

```
error: hidden lifetime parameters are deprecated, try `Foo<'_>`
 --> src/main.rs:5:5
  |
5 |     format!("foo {}", 22)
  |     ^^^^^^^^^^^^^^^^^^^^^
  |
note: lint level defined here
 --> src/main.rs:2:9
  |
2 | #![deny(elided_lifetime_in_path)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

I suppose the easiest fix is to fix the `format!` expansion, but maybe better to suppress this warning for macros outside of the current crate? Feels like a more general problem.

**Comment by nikomatsakis** [comment]
cc @dylan-dpc

**Comment by withoutboats** [comment]
> Feels like a more general problem.

I think the more general problem is that lints should be hygienic - we should only check them against ast nodes from this code, not those generated by a macro.

**Comment by shepmaster** [comment]
> we should only check them against ast nodes from this code

I've seen someone bit by something similar. They had a `deny(unsafe)` but a crate's macro ended up expanding to an `unsafe` block. IIRC, they fixed the library to add an `allow` on the block inside the macro.

So +1 for this kind of fix.

**Comment by durka** [comment]
That won't help with `forbid`.

**Comment by Dylan-DPC** [comment]
Any pointers on how to  suppress this warning for macros outside of the current crate?

**Comment by durka** [comment]
Clippy [has a function](https://github.com/rust-lang-nursery/rust-clippy/blob/master/clippy_lints/src/utils/mod.rs#L67-L95).

**Comment by nikomatsakis** [comment]
It does feel wrong to do this on a lint-by-lint basis, though, right? Maybe as part of `struct_lint_level`?

https://github.com/rust-lang/rust/blob/27a046e9338fb0455c33b13e8fe28da78212dedc/src/librustc/lint/mod.rs#L404-L420

We could e.g. return the "dummy" value for spans that are not in the current crate:

https://github.com/rust-lang/rust/blob/27a046e9338fb0455c33b13e8fe28da78212dedc/src/librustc/lint/mod.rs#L413

**Comment by Manishearth** [comment]
> It does feel wrong to do this on a lint-by-lint basis, though, right? Maybe as part of struct_lint_level?

**Comment by Manishearth** [comment]
> It does feel wrong to do this on a lint-by-lint basis, though, right? Maybe as part of struct_lint_level?

It's not that simple, though. The node being linted may be outside of a macro, but it may _contain_ macroy things that affect the lint. The correct way to handle this is to identify all the "important" nodes and do a macro check.

(Also: All currently supported ways of doing a macro check will not lint stuff like `some_macro!(some.code.that.hits.the.lint)`, which isn't an uncommon breed of macro.)

**Comment by nikomatsakis** [comment]
@Manishearth 

> The correct way to handle this is to identify all the "important" nodes and do a macro check.

So actually `struct_lint_level` is probably not the right place. Rather, `lint_level_at_node`:

https://github.com/rust-lang/rust/blob/27a046e9338fb0455c33b13e8fe28da78212dedc/src/librustc/ty/context.rs#L2222-L2247

That is, this seems like an identical problem to deciding whether the lint is enabled. We're basically talking about adding some measure of hygiene to that check, right?

(Also, I'm totally willing to be believe we should just work around this for now, and defer any more general changes to be done as part of a hygiene-overhaul-sort-of-thing.)

**Comment by nikomatsakis** [comment]
(One annoying problem: that fn does not presently have a span.)

**Comment by Manishearth** [comment]
It's _almost_ identical to that problem, and I think we can get away with that approximation for now.

I can think of counterexamples but it just makes the lints overconservative.

However, bear in mind that we may not want this _at all_ for future compat lints (we already catalog these so we can special case this).

**Comment by nikomatsakis** [comment]
@Manishearth 

> However, bear in mind that we may not want this at all for future compat lints (we already catalog these so we can special case this).

Interesting point. Probably we don't, yes. =) (The story around future compatibility and dependencies is sort of complicated, I think, but see also https://github.com/rust-lang/rust/issues/34596)

**Comment by Manishearth** [comment]
Issue for adding macro checks to lints at #48855

**Comment by reuvenpo** [comment]
This issue is also related to this topic: #51903.
Is this prioritised in any way? Seems like this had been known for a while now, and it's a shame because i would like to use this lint in my code.

**Comment by zackmdavis** [comment]
This is now fixed (lint was rewritten in #52069).

```
zmd@ReflectiveCoherence:~/Code/Misc$ cat 48385.rs 
#![feature(nll)]
#![deny(elided_lifetimes_in_paths)]

fn main() {
    format!("foo {}", 22);
}
zmd@ReflectiveCoherence:~/Code/Misc$ rustc +nightly --version
rustc 1.31.0-nightly (e7f5d4805 2018-10-18)
zmd@ReflectiveCoherence:~/Code/Misc$ rustc +nightly 48385.rs 
zmd@ReflectiveCoherence:~/Code/Misc$ echo $?
0
```
